### PR TITLE
Changes need to make akka-http compile with both 2.4 and 2.5

### PR DESCRIPTION
**Pull request by jrudolph**
Refs #956 and #957.

### CODE REVIEW

**Comment by jrudolph** [comment]
Removing `Fusing.aggressive` calls lead to a 15% performance hit using `ab` without persistent connections (3500 connections per second down to 3000 connections per second).

I therefore removed that change and replaced it with a change that uses reflection to call `Fusing.aggressive` in 2.4 to keep the old performance.

**Comment by jrudolph** [comment]
FWIW with the same code on 2.5 it's 7000 connections per second.

**Comment by ktoso** [comment]
> FWIW with the same code on 2.5 it's 7000 connections per second.

awww yiiis

**Comment by jrudolph** [comment]
PLS BUILD

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala [hunk 1]
```diff
@@ -35,7 +35,10 @@ private object PoolConductor {
         slotEventIn.carbonCopy(),
         slotOuts.map(_.carbonCopy()))
 
-    override def copyFromPorts(inlets: immutable.Seq[Inlet[_]], outlets: immutable.Seq[Outlet[_]]): Shape =
+    /**
+     * Needs to be implemented for 2.4. Not needed in 2.5 any more.
+     */
+    /* override */ def copyFromPorts(inlets: immutable.Seq[Inlet[_]], outlets: immutable.Seq[Outlet[_]]): Shape =
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala]: good, thanks for the comment

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala [hunk 2]
```diff
@@ -214,7 +214,7 @@ private[http] final class BodyPartParser(
                       emitFinalPartChunk: (List[HttpHeader], ContentType, ByteString) ⇒ Unit = {
                         (headers, ct, bytes) ⇒
                           emit(BodyPartStart(headers, { rest ⇒
-                            SubSource.kill(rest)
+                            StreamUtils.cancelSource(rest)(materializer)
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: ok good

File: akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala [hunk 3]
```diff
@@ -303,6 +224,52 @@ private[http] object StreamUtils {
...
+    case EmptySource ⇒ // nothing to do with empty source
+    case x ⇒
+      val mat =
+        GraphInterpreter.currentInterpreterOrNull match {
+          case null if materializer ne null ⇒ materializer
+          case null                         ⇒ throw new IllegalStateException("Need to pass materializer to cancelSource if not run from GraphInterpreter context.")
+          case x                            ⇒ x.subFusingMaterializer // try to use fuse if already running in interpreter context
+        }
+      x.runWith(Sink.ignore)(mat)
+  }
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala]: good

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala]: I wonder if my weird materializer could be used here actually once it's out there...

File: akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala [hunk 4]
```diff
@@ -303,6 +224,52 @@ private[http] object StreamUtils {
...
+
+      new Fuser {
+        def aggressive[S <: Shape, M](g: Graph[S, M]): Graph[S, M] = instance.aggressive(g)
+      }
+    } catch {
+      case ex: ClassNotFoundException ⇒
+        new Fuser {
+          def aggressive[S <: Shape, M](g: Graph[S, M]): Graph[S, M] = g // no fusing
+        }
+    }
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala]: ok good, allows us to work with both nicely for now

File: akka-http-core/src/test/scala/akka/stream/testkit/Utils.scala [hunk 5]
```diff
@@ -16,6 +16,11 @@ object Utils {
 
   case class TE(message: String) extends RuntimeException(message) with NoStackTrace
 
+  type ActorMaterializerImpl = {
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/stream/testkit/Utils.scala]: add a comment that this is for 2.4 compat?

**Pull request merged**